src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -209,18 +229,23 @@ private TReport RecordDiagnosticReport(Document document, VSDiagnostic[] diagnos
             }
         }
 
-        private static VSDiagnostic ConvertDiagnostic(Document document, SourceText text, DiagnosticData diagnosticData)
+        private VSDiagnostic ConvertDiagnostic(Document document, SourceText text, DiagnosticData diagnosticData)
         {
             Contract.ThrowIfNull(diagnosticData.Message, $"Got a document diagnostic that did not have a {nameof(diagnosticData.Message)}");
             Contract.ThrowIfNull(diagnosticData.DataLocation, $"Got a document diagnostic that did not have a {nameof(diagnosticData.DataLocation)}");
 
             var project = document.Project;
+
+            // Razor wants to handle all span mapping themselves.  So if we are in razor, return the raw doc spans, and
+            // do not map them.
+            var useMappedSpan = !document.IsRazorDocument();
```

Comment:
This should be "push" aware shouldn't it?

Reply:
this path is only ever used for 'pull'.  hence AbstractPullDiagnosticHandler.  Push goes through another path altogether.